src/api.rs

```diff
@@ -119,24 +116,40 @@ pub struct PlayersResponse {
     pub players: Vec<PlayerData>,
 }
 
+/// The current state for a player that is needed by the host site.
+///
+/// This doesn't include all of the player's internal state data, only the information needed
+/// by the display site.
+#[derive(Debug, Serialize)]
+pub struct PlayerData {
+    /// The player's ID.
+    id: PlayerId,
+
+    /// The player's display name.
+    username: String,
+
+    /// The player's current score.
+    score: usize,
+
+    /// The total number of balls in the players food pile.
+    balls: usize,
+}
+
 /// Returns a list of players and their scores.
 ///
 /// This is used by new host connections to update thier display to match the current state of the
 /// game.
 #[get("/players")]
-pub fn get_players(
-    scoreboard: State<Mutex<Scoreboard>>,
-    usernames: State<Mutex<Usernames>>,
-) -> JSON<PlayersResponse> {
-    // Clone the scoreboard and usernames table so we can work on their data without holding the
-    // mutex for too long.
-    let scoreboard = scoreboard.lock().expect("Scoreboard mutex was poisoned").clone();
-    let mut usernames = usernames.lock().expect("Usernames mutex was poisoned").clone();
-
-    let players = usernames.drain()
-        .map(|(id, username)| {
-            let score = *scoreboard.get(&id).expect("Score for player was not in scoreboard");
-            PlayerData { id, username, score }
+pub fn get_players(players: State<PlayerMap>) -> JSON<PlayersResponse> {
+    let players = players.read().expect("Player map was poisoned!");
```

Comment:
Before we were using a `Mutex`, now we're using `RwLock`. Mutex ensures that only one thread can access the data, and that thread always has mutable access once it has acquired the lock. `RwLock` has both `read()` method, which allows multiple threads to read the data at the same time, and a `write()` method, which behaves the same as `lock()` did.

Switching to `RwLock` give us a little bit of leeway to have multiple threads access the data in parallel if they don't need to mutate it.